### ISSUE TITLE
Epic: replicas are slow to start

### ISSUE THREAD

**Issue by kelvich** [neondatabase/neon#6006]
## Motivation

We observed following behavior on prod now:
1. primary actively writes the data, so there is some sk<>ps lag
2. ~~secondary checks last sk lsn~~ basebackup is asked with lsn=0
3. secondary tries to get basebackup with that lsn from ps
4. it takes more then 2m for pageserver to catch up to that lsn
5. secondary start times out (timeout is 2m) and we retry from step 2 and looping over

That lead to secondary start being stuck in that loop for a significant amount of time.

more details in https://neondb.slack.com/archives/C03F5SM1N02/p1701369162911829

## DoD

Replica is quick to start

## Implementation ideas

Most likely we don't have to start replica from `commmit_lsn`, we can start it from pageserver's `last_record_lsn`. Replica will start lagged, but that will happen anyway in a few minutes.

```[tasklist]
### Tasks
- [ ] validate proposed solution
- [ ] implement proposed solution : )
```

**Comment by kelvich** [comment]
cc @knizhnik @hlinnaka @ololobus -- do you think proposed fix makes sense?

**Comment by ololobus** [comment]
IIRC, previously if you don't provide LSN to basebackup it will create it from something that pageserver thinks as HEAD (likely last_record_lsn). Is it still true or some false memory? If it's true, we can simply skip sync safekeepers for ROs

> it takes more then 2m for pageserver to catch up to that lsn

Either way, can you please elaborate why it only affects ROs? @kelvich For RWs we also take HEAD from safekeepers perspective, which can be != last_record_lsn. One explanation I have in mind is that when we start RW, this means that there is no compute on this timeline, so the gap is likely small; when we start RO, it could be that there is a running and advancing compute, so the gap can be larger. Am I right?

It's still not clear why it's only RO problem. If the gap is huge, it likely affects all GetPage@LSN from RW, right?

**Comment by knizhnik** [comment]
I am not suer that we are trying to solve the real problem...
Why PS can not catch-up in more than 2 minutes?
We have quite small `max_replication_write_lag=15MB` and it should not take more than few seconds to reply this WAL.
Also not only replica can wait for commit LSN. Any `get_page` at this LSN can also be blocked by more than 2 minutes. But there is 1 minute wait for LSN timeout.

Sao IMHO the problem is not that we are trying to launch replica on wrong LSN, but that PS is not able to catch-up for so long time.  This is a real problem which needs to be investigated.  If it will be solved then no hack with spawning lagged replica is needed.

**Comment by kelvich** [comment]
we had another issue on that endpoint again. My initial suggestion in this ticket was wrong. For some reason basebackup is generally slow while there is active write on the primary. Lock contention?

**Comment by kelvich** [comment]
second thread: https://neondb.slack.com/archives/C03H1K0PGKH/p1701698080418069

**Comment by hlinnaka** [comment]
I spent some time looking at the logs from that time period, from different angles. I still don't understand what the root cause is. But a couple of suspicious / interesting / weird things caught my eye:

# Safekeeper connection keeps getting reset

The safekeeper connection keeps being disconnected and re-established: 
```
2023-12-04T15:04:41.344162Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(281), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-0.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1a"), reason: SwitchAvailabilityZone }
2023-12-04T15:05:04.871324Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(280), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-2.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1c"), reason: NoKeepAlives { last_keep_alive: Some(2023-12-04T15:04:47.373794003), check_time: 2023-12-04T15:05:04.871323163, threshold: 10s } }
2023-12-04T15:05:08.100344Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(281), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-0.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1a"), reason: SwitchAvailabilityZone }
2023-12-04T15:05:24.835772Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(280), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-2.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1c"), reason: NoKeepAlives { last_keep_alive: Some(2023-12-04T15:05:13.962288268), check_time: 2023-12-04T15:05:24.835770644, threshold: 10s } }
2023-12-04T15:05:36.493475Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(279), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-1.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1b"), reason: NoKeepAlives { last_keep_alive: Some(2023-12-04T15:05:25.803067141), check_time: 2023-12-04T15:05:36.493474415, threshold: 10s } }
2023-12-04T15:06:04.632385Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(281), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-0.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1a"), reason: SwitchAvailabilityZone }
2023-12-04T15:06:15.686945Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(279), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-1.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1b"), reason: NoKeepAlives { last_keep_alive: Some(2023-12-04T15:06:05.006008048), check_time: 2023-12-04T15:06:15.686942453, threshold: 10s } }
2023-12-04T15:07:06.050773Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(280), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-2.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1c"), reason: NoKeepAlives { last_keep_alive: Some(2023-12-04T15:06:55.956364997), check_time: 2023-12-04T15:07:06.050771639, threshold: 10s } }
2023-12-04T15:07:25.974980Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(279), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-1.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1b"), reason: NoKeepAlives { last_keep_alive: Some(2023-12-04T15:07:14.992612282), check_time: 2023-12-04T15:07:25.974977933, threshold: 10s } }
2023-12-04T15:07:39.927749Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(281), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-0.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1a"), reason: SwitchAvailabilityZone }
2023-12-04T15:08:01.674771Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(279), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-1.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1b"), reason: NoKeepAlives { last_keep_alive: Some(2023-12-04T15:07:44.097710608), check_time: 2023-12-04T15:08:01.674770530, threshold: 10s } }
2023-12-04T15:08:06.640602Z  INFO wal_connection_manager{tenant_id=0fbc345bd575026ae22758fc3e98af94 timeline_id=5e8a000bdd895528b37643eaad81cf37}: Switching to new connection candidate: NewWalConnectionCandidate { safekeeper_id: NodeId(281), wal_source_connconf: PgConnectionConfig { host: Domain("safekeeper-0.us-east-1.aws.neon.tech"), port: 6500, password: Some(REDACTED-STRING) }, availability_zone: Some("us-east-1a"), reason: SwitchAvailabilityZone }
```
 
Why is that? That shouldn't affect the basebackup, but weird..

## IP address is reused pretty quickly

Look at these two basebackup requests:

```
2023-12-04T15:07:25.529158Z ERROR page_service_conn_main{peer_addr=10.12.42.64:52806}: query handler for 'basebackup 0fbc345bd575026ae22758fc3e98af94 5e8a000bdd895528b37643eaad81cf37' failed: Broken pipe (os error 32)
Stack backtrace:
   ... <omitted> ...
2023-12-04T15:07:25.534627Z  INFO page_service_conn_main{peer_addr=10.12.42.64:52806}: Disconnected (Broken pipe (os error 32))
2023-12-04T15:07:53.784204Z  INFO page_service_conn_main{peer_addr=10.12.42.64:45288}:process_query{tenant_id=bdd616ef9e9c3c2cd6ceb591dda3c90e timeline_id=218aec985bc087d431828a52f0176f2b}:handle_basebackup_request{lsn=Some(0/1F20878) prev_lsn=None full_backup=false}: waiting for 0/1F20878
2023-12-04T15:07:53.785990Z  INFO page_service_conn_main{peer_addr=10.12.42.64:45288}:process_query{tenant_id=bdd616ef9e9c3c2cd6ceb591dda3c90e timeline_id=218aec985bc087d431828a52f0176f2b}:handle_basebackup_request{lsn=Some(0/1F20878) prev_lsn=None full_backup=false}: taking basebackup lsn=0/1F20878, prev_lsn=0/1F20850 (full_backup=false)
2023-12-04T15:07:53.807155Z  INFO page_service_conn_main{peer_addr=10.12.42.64:45288}:process_query{tenant_id=bdd616ef9e9c3c2cd6ceb591dda3c90e timeline_id=218aec985bc087d431828a52f0176f2b}:handle_basebackup_request{lsn=Some(0/1F20878) prev_lsn=None full_backup=false}:send_tarball{backup_lsn=0/1F20878}: timeline.pg_version 15
2023-12-04T15:07:54.580115Z  INFO page_service_conn_main{peer_addr=10.12.42.64:45288}:process_query{tenant_id=bdd616ef9e9c3c2cd6ceb591dda3c90e timeline_id=218aec985bc087d431828a52f0176f2b}:handle_basebackup_request{lsn=Some(0/1F20878) prev_lsn=None full_backup=false}: basebackup complete lsn_await_millis=0 basebackup_millis=795
```

They are coming from the same IP address, but are for completely different tenants. That's not wrong per se, it's normal for IP addresses to be reused. However, I'm surprised at how quickly it gets reused. I would've expect several minutes of cool-off in between.

I don't think this explains the basebackup problem. Unless there's some sort of a networking issue? But it seeems far-fetched.

**Comment by knizhnik** [comment]
> There is something strange here. All the log messages are printed in a small 20 ms window. But the last log line says lsn_await_millis=188243, implying that it took over 3 minutes. How come?

In theory it can happen if we block for 3 minutes in `get_active_tenant_timeline`.

**Comment by hlinnaka** [comment]
It's possible that this was fixed by https://github.com/neondatabase/neon/pull/6046. Next step is to wait until that is deployed, and monitor the logs to see if those errors still appear.

**Comment by vadim2404** [comment]
@hlinnaka , the PR you mentioned was released today in all regions

**Comment by hlinnaka** [comment]
I checked the production logs for this stack trace (with the `add_slru_segment`) again. The last instance was at 2023-12-21T11:57:53.358761Z, on us-east-2, but nothing since. Did the deployment happen before that?

I'm going to check the logs again in about 1 week. Maybe it still happens, but less frequently.

**Comment by vadim2404** [comment]
@hlinnaka, #6046 was a part of this release https://github.com/neondatabase/neon/pull/6090/commits  (Dec 11)

**Comment by vadim2404** [comment]
@hlinnaka, any updates?

**Comment by hlinnaka** [comment]
Checked the logs again. There are two new instances of that error in the logs, on 2024-01-10 and 2024-01-11.

Looking at the operations log for that project around the time of those errors, there was a _suspend_compute_ operation that failed and was retried multiple times, but eventually succeeded. At both of those errors. All start_compute operations  succeeded. That's weird; why would a suspend_compute cause a new basebackup to be taken from the pageserver?

**Comment by kelvich** [comment]
> That's weird; why would a suspend_compute cause a new basebackup to be taken from the pageserver?

we now do `--sync-safekeepers` at suspend to have faster starts. Could be that this is calling basebackup too by accident?

**Comment by andreasscherbaum** [comment]
@hlinnaka Where does this leave the ticket? Right now all the compute nodes succeed, but that does not confirm that the problem no longer exist, right?

**Comment by andreasscherbaum** [comment]
Should be fixed by changes from Konstantin. Closing this now, let's open new tickets if necessary.
